### PR TITLE
Make implementation of Faraday agnostic to data modules and features

### PR DESCRIPTION
**Pull request by shengy90**
Current Faraday source code is tightly coupled with the Data Module and cannot support an arbitrary data module with arbitrary features. For instance, model features were hard coded in FaradayVAE's training loop.

This PR decouples Faraday source codes and data modules:
- List of features are defined in the data modules which Faraday source could will infer from
- Faraday VAE and GMM modules can arbitrary train on datasets with any prescribed features, instead of being hard coded to "month" and "dayofweek" as it is currently
- This will require data modules to return: two items 1) the kWh and 2) a dictionary of feature tensors. As long as the data modules satisfy these 2 criteria, Faraday's source could will work accordingly

~TODO:~
- Write tests ✅ 
- Test code with Non-LCL dataset ✅

### CODE REVIEW

**Comment by charlotte-avery** [comment]
I think it would be helpful to add a note to the README outlining the data module requirements (i.e., _return: two items 1) the kWh and 2) a dictionary of feature tensors_) so that if a user wants to write their own data module (for their own dataset) they understand how to set it up.

File: src/opensynth/models/faraday/model.py [hunk 1]
```diff
@@ -372,26 +477,29 @@ def sample_gmm(
...
         gmm_kwh = gmm_samples[:, : self.vae_module.latent_dim]
-        gmm_mth = np.round(gmm_samples[:, -2], decimals=0).astype(int)
-        gmm_dow = np.round(gmm_samples[:, -1], decimals=0).astype(int)
+        gmm_labels: dict[str, torch.tensor] = {}
+        # Abstract the labels and round numerical values to integers
+        for i, feature in enumerate(self.vae_module.feature_list):
+            index = self.get_index(self.vae_module.feature_list, i)
+            gmm_labels[feature] = np.round(
+                gmm_samples[:, index], decimals=0
+            ).astype(int)
```
Comment (charlotte-avery) [inline src/opensynth/models/faraday/model.py:486]: Should all features be converted to `int` types? Could it be possible that the user intended other input features to be of a different datatype? Or is `int` a requirement of features?

File: src/opensynth/models/faraday/model.py [hunk 2]
```diff
@@ -329,6 +345,90 @@ def __init__(
...
+
+        Args:
+            feature_list (list[str]): List of features
+            current_index (int): Current iter index
+
+        Returns:
+            int: Returns colum index
+        """
+        return -(len(feature_list) - current_index)
+
```
Comment (charlotte-avery) [inline src/opensynth/models/faraday/model.py:431]: I feel that it would be cleaner to have these static modules (`get_feature_range`, `create_mask`, `get_index`) taken out of FaradayModel into their own class, given that they have a specific purpose of filtering out invalid samples.

File: src/opensynth/models/faraday/model.py (hunk 1 above)
Comment (charlotte-avery) [inline src/opensynth/models/faraday/model.py:486]: I am not sure I understand why this code has to be in a for-loop? 
From my understanding, `get_index` just converts indices [0,1,2,3] -> [-4,-3,-2,-1] s.t. it can access the features in `gmm_samples`, so I propose `get_index` returns an array (or tensor), r.t. an integer, given by `-(len(self.vae_module.feature_list) - range(0, len(self.vae_module.feature_list), 1)[::-1]` . This avoids the need for a for-loop.

File: src/opensynth/models/faraday/model.py (hunk 2 above)
Comment (charlotte-avery) [inline src/opensynth/models/faraday/model.py:431]: From the docstring, it is not entirely clear why `get_index` is necessary. From my understanding, it is to find the location of the feature in `samples`.

File: src/opensynth/models/faraday/model.py (hunk 1 above)
Comment (charlotte-avery) [reply src/opensynth/models/faraday/model.py:486]: Also the rounding step can be taken out the for-loop (assuming that you can do `np.round(gmm_samples[:, list_indicies], decimals=0).astype(int)`)

File: tests/models/faraday/test_faraday_model.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+@pytest.fixture
+def feature_dict() -> dict[str, torch.tensor]:
+    feature_list = ["feature_1", "feature_2"]
+    output_dict: dict[str, torch.tensor] = {}
+    for feature in feature_list:
+        # numpy randint high is exclusive!
+        random_tensor = np.random.randint(low=0, high=6, size=100)
+        output_dict[feature] = torch.from_numpy(random_tensor)
+    return output_dict
```
Comment (charlotte-avery) [inline tests/models/faraday/test_faraday_model.py:22]: What happens if the feature is not an`int`?

File: src/opensynth/models/faraday/model.py (hunk 1 above)
Comment (shengy90) [reply src/opensynth/models/faraday/model.py:486]: Theoretically no, GMM by default handle floats, but it's a bit cumbersome to have to cater for non-integer encoded features. I'll add this to docstrings and end users can always subclass `FaradayModel` and overwrite this method to support float labels if they so wish.

File: tests/models/faraday/test_faraday_model.py (hunk 3 above)
Comment (shengy90) [reply tests/models/faraday/test_faraday_model.py:22]: Not support for now - end users should subclass FaradayModel and overwrite sample_gmm/ write their own tests for custom implementation if they wish to support non-integer labels.

Will leave implementing non-integer labels into OpenSynth directly if we have a need for it ourselves.

File: src/opensynth/models/faraday/model.py (hunk 2 above)
Comment (shengy90) [reply src/opensynth/models/faraday/model.py:431]: But these functions are also specific to `FaradayModel` only though, and doesn't make sense as a standalone. Taking it out makes this out of context.

We can always take this out into a separate utils.py if other classes of model require this.

Comment (shengy90) [reply src/opensynth/models/faraday/model.py:431]: I'm writing it as a function so that it can be unit tested.

Comment (charlotte-avery) [reply src/opensynth/models/faraday/model.py:431]: Yes, I understand, can we update the docstring to clarify why this function is necessary e.g., "extract feature column locations in gmm_samples"

File: src/opensynth/models/faraday/model.py (hunk 1 above)
Comment (shengy90) [reply src/opensynth/models/faraday/model.py:495]: It's not just rounding, but also appending to the `gmm_labels` dictionary in an ordered manner, so that it can be decoded by the VAE.  The order of the dictionary needs to be preserved, and that check is handled in the `train_gmm` step (to check that the data module used to train the GMM module loads in features in the same order as seen by VAE). Otherwise you're sampling/ conditioning on the wrong labels. 

So TL;DR the for loop can't be avoided. It needs to iterate through the label columns so that it can be stored in a python dictionary in the right order. And if the for-loop can't be avoided, I don't quite see the value of taking the rounding step out of the for-loop.

The alternative implementation of taking the the rounding out of the for-loop without the use of `get_index` as you suggested would be:
```python
  gmm_kwh: torch.Tensor = gmm_samples[:, : self.vae_module.latent_dim]

  gmm_label_int: npt.NDArray[np.int64] = np.round(
      gmm_samples[:, self.vae_module.latent_dim :], decimals=0
  )
  gmm_labels: dict[str, torch.Tensor] = {}

  for i, feature in enumerate(self.vae_module.feature_list):
      gmm_labels[feature] = gmm_label_int[:, i]
```

But I don't think this implementation is solving anything and in fact it makes the code less explicit as the index from `for i, feature in enumerate(self.vae_module.feature_list):` is no longer coupled explicitly with the index of `gmm_label_int` array. To unit test the the logic, we then have to wrap this in a much larger function and test more things.

It's more unit testable I think to abstract the logic in the `get_index` function so we can just test it explicitly. There's a bug with the pytest however - fixing in latest commit.

**Comment by shengy90** [comment]
> 
I think it would be helpful to add a note to the README outlining the data module requirements (i.e., return: two items 1) the kWh and 2) a dictionary of feature tensors) so that if a user wants to write their own data module (for their own dataset) they understand how to set it up.


This is already typed as `def forward(self, input_data: TrainingData):` though, so I think that's already self-explanatory.

File: src/opensynth/models/faraday/model.py (hunk 2 above)
Comment (shengy90) [reply src/opensynth/models/faraday/model.py:439]: Have update docstring in latest commit.

File: src/opensynth/models/faraday/model.py (hunk 1 above)
Comment (charlotte-avery) [reply src/opensynth/models/faraday/model.py:497]: Okay, I can see why the loop is simpler/clearer/easier for testing. Thanks for the response. Ideally, should this for loop also be in a function so it can be unit tested to ensure `gmm_labels` are as expected?

**Review by charlotte-avery — APPROVED** [review]
LGTM. You could also add a comment in the `TrainingData` docstring to flag that `features` are expected to be integers, for clarity for the user.

File: src/opensynth/models/faraday/model.py (hunk 1 above)
Comment (shengy90) [reply src/opensynth/models/faraday/model.py:497]: We could put the for loop too in a function to test it, but depends on how we are writing the test for it!
